### ISSUE TITLE
Make chai.js and SinonJS compatible with testing framework and add instructions how to use it

### ISSUE THREAD

**Issue by bartlomieju** [denoland/deno_std#1284]
[`chai.js`](https://www.chaijs.com/) and [`Sinon.JS`](https://sinonjs.org/) are very popular testing libraries that run both in Node and browser.

Since they can be run in browser it should be a low hanging fruit to make them work in Deno and I believe it would be very valuable to add "first-class" instructions to `testing/README.md` how they could be used with Deno's built-in testing (`deno test`).

CC @keithamus

**Comment by keithamus** [comment]
There's some slow progress to getting chai working in Deno. We're working on migrating the dependencies right now so that they work "natively" in Deno & Node. I hope to include TypeScript support also.

**Comment by bartlomieju** [comment]
> There's some slow progress to getting chai working in Deno. We're working on migrating the dependencies right now so that they work "natively" in Deno & Node. I hope to include TypeScript support also.

Thanks for the heads up, is there anything I can assist with? Are you facing any challenges to target both runtimes?

**Comment by bartlomieju** [comment]
> > Are you facing any challenges...
> 
> Mostly motivation 😂. Deno is pleasurable to work with although it can be prickly with regard to imports and the `.ts` extension which Node+TypeScript doesn't like, but Deno requires. I'm not sure if there is good tooling to help manage the difference in dependency APIs?

Keith I noticed that `chai` already has [browser support](https://www.chaijs.com/guide/installation/#browser); I'm willing to guess that it should just work in Deno without any changes. Maybe the only thing you want to do is link that output to https://deno.land/x/ to make it more discoverable for Deno users?

**Comment by cmorten** [comment]
@bartlomieju I believe you are right, e.g.

```ts
// @deno-types="https://unpkg.com/@types/chai@4.2.22/index.d.ts"
import "https://chaijs.com/chai.js";

Deno.test("we can make chai assertions", () => {
  const assert = chai.assert;
  const foo = "bar";
  const beverages = { tea: ["chai", "matcha", "oolong"] };

  assert.typeOf(foo, "string"); // without optional message
  assert.typeOf(foo, "string", "foo is a string"); // with optional message
  assert.equal(foo, "bar", "foo equal `bar`");
  assert.lengthOf(foo, 3, "foo`s value has a length of 3");
  assert.lengthOf(beverages.tea, 3, "beverages has 3 types of tea");
});

Deno.test("we can make chai expectations", () => {
  const expect = chai.expect;
  const foo = "bar";
  const beverages = { tea: ["chai", "matcha", "oolong"] };

  expect(foo).to.be.a('string');
  expect(foo).to.equal('bar');
  expect(foo).to.have.lengthOf(3);
  expect(beverages).to.have.property('tea').with.lengthOf(3);
});

Deno.test("we can use chai should style", () => {
  chai.should();
  const foo = "bar";
  const beverages = { tea: ["chai", "matcha", "oolong"] };

  foo.should.be.a('string');
  foo.should.equal('bar');
  foo.should.have.lengthOf(3);
  beverages.should.have.property('tea').with.lengthOf(3);
});
```

works just fine with `deno test ./testing/chai_example.ts`.

Similarly chai seems to work nicely as well, e.g.

```ts
import sinon from "https://cdn.skypack.dev/sinon@11.1.2?dts";
import chai from "https://cdn.skypack.dev/chai@4.3.4?dts";

Deno.test("stubbed callback", () => {
  const callback = sinon.stub();
  callback.withArgs(42).returns(1);
  callback.withArgs(1).throws(new Error("test-error"));

  chai.assert.isUndefined(callback()); // No return value, no exception
  chai.assert.equal(callback(42), 1); // Returns 1
  chai.assert.equal(callback.withArgs(42).callCount, 1); // Use withArgs in assertion
  chai.assert.throws(() => {
    callback(1);
  }, "test-error"); // Throws Error("test-error")
});
```

with `deno test ./testing/sinon_example.ts`.

Also assume findings from a while ago mocha should also work (REF: https://dev.to/craigmorten/testing-your-deno-apps-with-mocha-4f35)

**Comment by cmorten** [comment]
As a starter for 10 have thrown these examples into a PR https://github.com/denoland/deno_std/pull/1334 - sure can be used as the basis of a README update somewhere as well.

**Comment by kt3k** [comment]
addressed in #1334. Thank you @cmorten for the work!
